ea-2022-04-ndvi-automation.ipynb

```diff
@@ -0,0 +1,1600 @@
+    "\n",
+    "    all_masked_values = cloud_shadow + cloud + high_cloud_confidence\n",
+    "    cl_mask = landsat_qa.isin(all_masked_values)\n",
+    "\n",
+    "    return cl_mask\n",
+    "\n",
+    "\n",
+    "def combine_scenes(path_to_all_scenes, crop_path):\n",
+    "    \"\"\"Open, cleans, crops list of raster files, cleans data with cloud mask, \n",
+    "    captures average ndvi, date, and site name and saves to pandas df. \n",
+    "\n",
+    "    Parameters\n",
+    "    ----------\n",
+    "    path_to_all_scenes: string\n",
+    "        Path to directory with all landsat scenes/date folders to be combined.\n",
+    "    crop_path: string\n",
+    "        A string path to the geodataframe containing the crop geometry that \n",
+    "        you wish to crop your data to.\n",
+    "\n",
+    "    Returns\n",
+    "    ----------\n",
+    "    df : pandas dataframe\n",
+    "        Dataframe containing average NDVI values, date, and site name. \n",
+    "        \"\"\"\n",
+    "\n",
+    "    dates_temp = []\n",
+    "    ndvi_values_temp = []\n",
+    "\n",
+    "    for ascene in path_to_all_scenes:\n",
+    "        paths_list = sorted(glob(os.path.join(ascene, \"*band[4-5]*.tif\")))\n",
+    "        landsat_scene = process_bands(paths_list, crop_path, stack=True)\n",
+    "\n",
+    "        # Mask data with pixel QA layer\n",
+    "        qa_path = glob(os.path.join(ascene, '*pixel_qa.tif'))\n",
+    "        landsat_scene_cl_free = landsat_scene.where(\n",
+    "            ~remove_clouds(qa_path[0]))\n",
+    "\n",
+    "        landsat_scene_ndvi = normalized_diff(\n",
+    "            landsat_scene_cl_free[1], landsat_scene_cl_free[0])\n",
+    "\n",
+    "        ndvi_values_temp.append(landsat_scene_ndvi)\n",
+    "\n",
+    "        date = ascene[50:58]\n",
+    "        datetime_object = datetime.strptime(date, '%Y%m%d')\n",
+    "        dates_temp.append(datetime_object)\n",
+    "\n",
+    "        df = pd.DataFrame({'mean_ndvi': ndvi_values_temp}, index=dates_temp)\n",
+    "        df['site'] = ascene[22:26]\n",
+    "\n",
+    "    df['mean_ndvi'] = df['mean_ndvi'].apply(pd.to_numeric, errors='coerce')\n",
+    "    df.drop(df[df['mean_ndvi'] == 0].index, inplace=True)\n",
+    "\n",
+    "    return df"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Using Functions for Optimized Code \n",
+    "These functions defined above help to make my notebook code more efficient and optimized. Instead of repeating several lines of code for every time I need to calculate NDVI, open/clean/crop a list of raster files, remove cloud interference from landsat data, or combine NDVI values for several sites/scenes into a pandas dataframe, I can use the functions defined above. These functions reduce the potential for error, make it easier to add more data to this analysis, and help someone else understand my code better. "
```

Comment:
Great use of multiple functions to complete the workflow! It is possible to combine some steps in the above functions into two functions to complete the workflow using for loops, but using multiple functions not only properly completes the required tasks, but adds an element of customization to future workflows that would not necessarily be possible with fewer functions. Great job!

Reply:
Thank you! I actually like the readability of your method (using fewer, more comprehensive functions with for loops) better. But this method did help me break the process down into smaller, more digestible pieces, which is what I needed in order to understand how to analyze data for all the sites/dates.